src/metabase/models/field_values.clj

```diff
@@ -138,28 +153,42 @@
 
 (t2/define-before-insert :model/FieldValues
   [{:keys [field_id] :as field-values}]
-  (u/prog1 (merge {:type :full}
-                  field-values)
+  (u/prog1 (merge {:type :full} field-values)
     (assert-valid-human-readable-values field-values)
-    (assert-valid-field-values-type field-values)
-    ;; if inserting a new full fieldvalues, make sure all the advanced field-values of this field is deleted
-    (when (= (:type <>) :full)
+    (assert-valid-type-hash-combo field-values)
+    ;; if inserting a new full fieldvalues, make sure all the advanced field-values of this field are deleted
+    (when (= :full (keyword (:type <>)))
       (clear-advanced-field-values-for-field! field_id))))
 
 (t2/define-before-update :model/FieldValues
   [field-values]
-  (let [{:keys [type values hash_key]} (t2/changes field-values)]
-    (u/prog1 field-values
-      (assert-valid-human-readable-values field-values)
-      (when (or type hash_key)
-        (throw (ex-info (tru "Can't update type or hash_key for a FieldValues.")
-                        {:type        type
-                         :hash_key    hash_key
-                         :status-code 400})))
-      ;; if we're updating the values of a Full FieldValues, delete all Advanced FieldValues of this field
-      (when (and values
-                 (= (:type field-values) :full))
-        (clear-advanced-field-values-for-field! (:field_id field-values))))))
+  (u/prog1 field-values
+    (assert-no-identity-changes field-values)
+    (assert-valid-human-readable-values field-values)
+    ;; if we're updating the values of a Full FieldValues, delete all Advanced FieldValues of this field
+    (when (and (contains? field-values :values) (= :full (keyword (:type field-values))))
+      (clear-advanced-field-values-for-field! (:field_id field-values)))))
+
+(defn- assert-coherent-query [{:keys [type hash_key] :or {type :full} :as field-values}]
+  (if (= :full (keyword type))
+    (when (some? hash_key)
+      (throw (ex-info "Invalid query - :full FieldValues cannot have a hash_key"
+                      {:field-values field-values})))
+    (when (and (contains? field-values :hash_key)
+               (nil? hash_key))
+      (throw (ex-info "Invalid query - Advanced FieldValues must have a hash_key"
+                      {:field-values field-values})))))
```

Comment:
This logic is duplicated on https://github.com/metabase/metabase/pull/38751/files#diff-b28eaae7ad5ee9389359e276833936958b869f4e239e8b2c4adb76078562bed9R116-R127

perhaps `assert-valid-type-hash-combo` should be used in the `t2/define-before-select` as well?

Reply:
I initially went that route of reusing it, but eventually decided against the guard that Cam suggested and you referenced, will discuss further on your summary comment.

Reply:
Oh, I remembered another reason I felt uncomfortable reusing it, the format of the exceptions. Having a translated error message and status code felt strange for an error that can be triggered outside of web requests. I really dislike how we're peppering presentation logic in the model layer - violating two layers of abstraction 😢

Reply:
Oh right. In that case I think `assert-valid-type-hash-combo` can be replaced by `assert-coherent-query`. I don't see why we need `status-code` and `tru` at this level, so maybe it's a mistake to leave them here. If the app throws these errors during an API request it's definitely unexpected and not an issue with an API call.

Reply:
I don't see how `field/:id/values` can hit it, but fair point on not wanting to introduce behaviour changes. I still think there's a way to dedupe or couple them explicitly

Reply:
Just to make sure I'm not missing something silly - that API end-point can hit insert, which will hit `assert-valid-type-hash-combo`, and throw those status codes. I didn't see anything that would catch and translate them within the handler.

Reply:
Ok, you're talking about this right? 

https://github.com/metabase/metabase/blob/4d552f131f509ad1aacaea379edb5c7fdc699601/src/metabase/api/field.clj#L323

`update-map` can only contain `:values` and `:human_readable_values`.

Then `create-field-values!` adds `type=full` and the `field_id`:
https://github.com/metabase/metabase/blob/4d552f131f509ad1aacaea379edb5c7fdc699601/src/metabase/api/field.clj#L304

`assert-valid-type-hash-combo` will never throw under these conditions.

Reply:
I was talking about the GET, which follows a slightly more tangled path and ends up somewhere different. In both cases the low level method hard codes the type and can't be passed a hash key. All I was say is that we reach the code, not that the current path is able to pass bad data. 

In practice I don't know if any APIs can trigger the exception, which I see was your point. My stance is that the scope is the contract, as finer details can be missed or change. I don't know anything about how our middleware or frontend handle errors yet, so I'm an extreme wimp in this area.

Reply:
Right yeah I was talking about in practice - whether we would be changing any observable behaviour of the BE by removing them.

Anyway, what I'd suggest here:
1. leave a FIXME comment or something in `assert-valid-type-hash-combo` because we potentially found some code that could be deleted
2. I was torn about deduping `assert-valid-type-hash-combo` and `assert-coherent-query` for the reasons you said (why should select and insert be coupled like this), but I think they do contain some duplicated logic: `type` and `hash_key` and are coupled components of a unique identity for an advanced field values row. You shouldn't be able to select or insert part of the identity (e.g. only `hash_key`). So `assert-valid-type-hash-combo` is probably the correct name for such a function that checks the identity is whole, and it should be used in both the select and insert hook. I think it should be wrapped in a `(when type ...` too, as it was before.

Reply:
Looking closer, I can't de-duplicate the logic because it's not quite the same right now, although perhaps they will be if select validation is made more opinionated. The difference right now is that advanced field values query can omit the hash_key.

Reply:
Right. What's the downside of making select validation require the `hash_key` if `type!=full`?

Reply:
But you're right, I guess it's fair to leave it deduped. I'm not strongly convinced they're inherently coupled